File: src/main/java/de/pflugradts/pwman3/application/commandhandling/handler/GetCommandHandler.java
package de.pflugradts.pwman3.application.commandhandling.handler;

import com.google.common.eventbus.Subscribe;
import com.google.inject.Inject;
import de.pflugradts.pwman3.application.ClipboardAdapterPort;
import de.pflugradts.pwman3.application.UserInterfaceAdapterPort;
import de.pflugradts.pwman3.application.commandhandling.command.GetCommand;
import de.pflugradts.pwman3.domain.model.transfer.Bytes;
import de.pflugradts.pwman3.domain.model.transfer.Output;
import de.pflugradts.pwman3.domain.service.password.PasswordService;

public class GetCommandHandler implements CommandHandler {

    @Inject
    private UserInterfaceAdapterPort userInterfaceAdapterPort;
    @Inject
    private ClipboardAdapterPort clipboardAdapterPort;
    @Inject
    private PasswordService passwordService;

    @Subscribe
    private void handleGetCommand(final GetCommand getCommand) {
        passwordService.viewPassword(getCommand.getArgument()).ifPresent(
            passwordBytes -> clipboardAdapterPort.post(Output.of(passwordBytes))
        );
        userInterfaceAdapterPort.send(Output.of(Bytes.of("Password copied to clipboard.")));
        getCommand.invalidateInput();
        userInterfaceAdapterPort.sendLineBreak();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/commandhandling/handler/egg/GetCommandHandler.kt
package de.pflugradts.passbird.application.commandhandling.handler.egg
import de.pflugradts.passbird.application.ClipboardAdapterPort
import de.pflugradts.passbird.application.UserInterfaceAdapterPort
import de.pflugradts.passbird.application.commandhandling.CommandExecutionTracker
import de.pflugradts.passbird.application.commandhandling.command.GetCommand
import de.pflugradts.passbird.application.commandhandling.handler.TypedCommandHandler
import de.pflugradts.passbird.application.useScrambled
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.outputOf
import de.pflugradts.passbird.domain.service.password.PasswordService
class GetCommandHandler constructor(
    private val passwordService: PasswordService,
    private val clipboardAdapterPort: ClipboardAdapterPort,
    private val userInterfaceAdapterPort: UserInterfaceAdapterPort,
    private val commandExecutionTracker: CommandExecutionTracker,
) : TypedCommandHandler<GetCommand>(GetCommand::class.java) {
    override fun handleCommand(command: GetCommand) {
        passwordService.viewPassword(command.argument).ifPresentOrElse(
            block = { password ->
                password.useScrambled {
                    val clipboardResult = clipboardAdapterPort.post(outputOf(it))
                    if (clipboardResult.failure) {
                        commandExecutionTracker.markFailure()
                    }
                    clipboardResult.onSuccess {
                        userInterfaceAdapterPort.send(outputOf(shellOf("Password copied to clipboard.")))
                    }
                }
            },
            other = {
                commandExecutionTracker.markFailure()
            },
        )
        command.invalidateInput()
        userInterfaceAdapterPort.sendLineBreak()
    }
}
